### ISSUE TITLE
pg_repack fails when rebuild deferred unique constraint

### ISSUE THREAD

**Issue by averkol** [reorg/pg_repack#209]
I have a table with an unique constraint in deferred mode. An application inserts and updates several records during one transaction and these records can violent this constraint until transaction is finished (all records are inserted/updated). It seemed that pg_repack can't handle this situation.
Example to reproduce:
```
CREATE TABLE test_table (
  id   SERIAL PRIMARY KEY,
  val1 INTEGER,
  val2 INTEGER
);
ALTER TABLE test_table ADD CONSTRAINT uk_test_table UNIQUE (val1, val2) DEFERRABLE INITIALLY DEFERRED;
INSERT INTO test_table (val1, val2) VALUES (1, 0);

DO
$$
DECLARE
  i INTEGER;
  v_id INTEGER;
BEGIN
  FOR i IN 1..10000 LOOP
    BEGIN
      INSERT INTO test_table (val1, val2) VALUES (1, 0) RETURNING id INTO v_id;
      UPDATE test_table set val2 = i where id = v_id;
      RAISE notice 'insert %', i;
      COMMIT;
    END;
  END LOOP;
END $$;
```
```
./pg_repack -t test_table -- when loop is running
```
RESULT:
```
INFO: repacking table "public.test_table"
ERROR: query failed: ERROR:  duplicate key value violates unique constraint "index_1280053"
DETAIL:  Key (val1, val2)=(1, 0) already exists.
CONTEXT:  SQL statement "INSERT INTO repack.table_1280047 VALUES ($1.*)"
DETAIL: query was: SELECT repack.repack_apply($1, $2, $3, $4, $5, $6)
```
I assume that the reason is that pg_repack creates a copy of the constraint without the deferred flag, so the constraint is checked straight after trigger re-inserts a new record.

**Comment by dvarrazzo** [comment]
Or maybe that it doesn't work into a transaction.

The easiest way to solve the problem is that you drop the constraint.

**Comment by averkol** [comment]
Unfortunately, I can't do this. I need the uniqueness of these attributes, and without the constraint, obviously, the uniqueness will be corrupted easily in situation of concurrent modifications of partly same records by several transactions.

**Comment by dvarrazzo** [comment]
Do you have a suggestion about what pg_repack should do?

**Comment by averkol** [comment]
Well, I found out that for a table with an unique constraint pg_repack creates a copy of the table without any constraints and also creates an unique index. But result is a bit different in the case when the original  constaint is deferred.
I think pg_repack should create a copy of the table with same unique constraints instead of creating indexes.

**Comment by dvarrazzo** [comment]
Unique constraints are implemented using an index, you can't implement an unique constraint without an index.

I don't think pg_repack can work in your use case. If you feel that's not the case, a patch is welcome.

**Comment by averkol** [comment]
> Unique constraints are implemented using an index, you can't implement an unique constraint without an index.

Yes, but I mean that there is a difference between creating unique constraint (that also leads to the creation of an index) and creating only unique index [(index-unique-checks)](https://www.postgresql.org/docs/11/index-unique-checks.html)

> If you feel that's not the case, a patch is welcome.

Ok, I don't have a lot of experience with programming in C, but, probably I'll try :-)

**Comment by NikolayS** [comment]
@averkol it definitely looks like a pg_repack's bug (or maybe a missing feature).

There is a trick that might help you here.

Internally, pg_repack creates indexes using definition provided by internal `repack_indexdef` function. For your example:

```
test=# select repack.repack_indexdef(indexrelid, indrelid, null, true) from pg_index where indexrelid = (select oid from pg_class where relname = 'uk_test_table');
                                                   repack_indexdef
---------------------------------------------------------------------------------------------------------------------
 CREATE UNIQUE INDEX CONCURRENTLY index_74555953 ON public.test_table USING btree (val1, val2) TABLESPACE pg_default
(1 row)
```

You can try the following approach. It doesn't require C coding at all, but some plpgsql code will be needed.

1) After installing pg_repack, rename its `repack_indexdef` function:
```
test=# alter function repack.repack_indexdef(oid, oid, name, boolean) rename to repack_indexdef_orig;
ALTER FUNCTION
```
2) Define your own `repack_indexdef` function in `repack` schema, with the same arguments, written in plpgsql (or just regular SQL, or plpython or anything you prefer). Inside it call the original function, `repack_indexdef_orig`, and analyze its output. Modify the output for particular objects (based on table/index names or OIDs), rewriting definition from `CREATE UNIQUE INDEX` to `ALTER TABLE .. ADD CONSTRAINT` with the desired flags `DEFERRABLE INITIALLY DEFERRED`.

Hope it will be helpful for you.

Of course, it doesn't mean that the problem shouldn't be fixed inside pg_repack itself.

**Comment by averkol** [comment]
@NikolayS Thank you for the idea. I'll try your workaround.
